### DISCUSSION TITLE
.Net MAUI seems to be a dead end, alternatives?

### DISCUSSION THREAD

**Discussion by MBrekhof** [dotnet/maui#10175] [General]
As a (former) certified Xamarin developer having tried to build a rather simple .NET MAUI mobile app I have the distinct impression it is not now usable. Question is whether it will ever be usable/stable enough or that it goes the way of the dodo. Which begs the question which alternatives are there: flutter or UNO come to mind. Any ideas?

**Comment by JohnHDev** [comment]
I don't have an answer for you, but some relatively good news, Xamarin Forms support has been extended to 2024, so you might not need to migrate or make a decision immediately.
I don't think Maui is dead, but we have paused most of our work on it until there is significant improvement. We will reevaluate in a few months and then start making plans at that point.

**Comment by AlleSchonWeg** [comment]
XF support to 2024 😂😂😂
Perhaps on the paper.

**Comment by akhanalcs** [comment]
Avalonia is a good alternative IMO. 
Never liked UNO.

**Comment by AlistairMcKinnon** [comment]
Try it again, it suddenly coming together...

**Comment by MarkusRodler** [comment]
Blazor Maui is ok if it has to be a native App. Otherwise Blazor SPA could also be an option.

**Comment by azrinsani** [comment]
MAUI is still very bad as of 2024. Give it another 3 years before it stabilizes.

I have started to port all my existing apps to Flutter.

**Comment by Stedy59** [comment]
Maui is pretty good as of 2024...

**Reply by ajay-mundi** [reply]
Are you on .net 8? Considering moving over from .net 7 but just a little apprehensive about the stability on .net 8

**Reply by JohnHDev** [reply]
Well, .NET 8 received all the bug fixes whereas .NET 7 only received some of them, if they were easy enough to back port apparently. We have been waiting for the .NET 8 release before attempting to migrate to .NET Maui. So far, it is looking much more stable, and certainly faster app start times on release builds.

**Reply by AlistairMcKinnon** [reply]
I've moved to .NET 8 as maui is relatively new so will have bug fixes and enhancements. Seems pretty good so far.

**Reply by ghost** [reply]
Upgraded my MAUI project from .NET 7 to .NET 8 because some things stopped working in .NET 7 but after the upgrade, it just got worse. Slower even. Goodbye smooth page transitions and animated popups. Come to think about it, .NET 8 is being advertised with improved performance. I really wonder on what OS/platform that is. I'd personally consider working on the .NET 7 version of your project until some more fixes have been released.

**Reply by zleao** [reply]
My experience going from .Net 7 to .Net 8 was very positive. Bugs were fixed and the overall developer experience as improved.
The negative part I found in .Net MAUI is that it feels I'm going through all the pain again of learning what are the perks and details for a new platform. I started to work with Xamarin back in 2013 when it was only Xamarin Native (no Xamarin Forms yet) and during all these years a lot of learning was needed, specially for the small detail stuff between platforms (iOS and Android).
The 3rd party support for MAUI is also very limited when compared to what exists for Xamarin. and it relies a lot on open source projects that are in 99% of the cases, maintained on good will.
The paid libraries are in most cases just very expensive and I've found that is not easy to choose one... Syncfusion is a nice one. I've also worked with Telerik in the past and it also had a very complete set of controls.
The GrialUI toolkit I haven't tried. Maybe some of you have some insights on this..?

Another negative point for me was the deprecation of VisualStudio for Mac. Unfortunatelly the experience to develop for iOS, connection VS for PC to a Mac is far from being acceptable. For me there's always conneciton problems and a poor developer experience. When I develop directly on a Mac it's much smoother. I Tried VisualStudio code with the MAUI extensions, but it's just to bad to even try to adapt to it. I'm now using Rider, and the experience is much better. In time maybe some nice extensions will come to Rider to give better support to MAUI...

**Comment by AlistairMcKinnon** [comment]
Yes, I have been through the same pain, and have been on the point of giving up.
The one thing I do now is back up the source code far more than before, particularly if doing anything with Nuget. Sometimes just updating with Nuget renders the whole thing completely unusable and unrecoverable.
I also create more small test apps than before, then update all the Nuget stuff, or add what I need, then test, then repeat for the main apps.
Sometimes you can have a spelling mistake in the code, and VS adds a “using” line which causes untold errors, which is a pain.
I have looked at Uno which seems pretty good and also addresses the Web app problem that Maui doesn’t for some reason but don’t like the way you need to learn a new XAML format.
Good luck!

**Reply by MarkusRodler** [reply]
I can recommend using the Blazor Variant. It is far more stable than the Xaml-Variant.

**Reply by JLfub** [reply]
If you don't need to work with any webviews, that is

**Comment by gautambjain** [comment]
I started migrating my Xamarin Forms project to .Net Maui (.Net 8) in March 2024 and I am enjoying the journey. I haven't had any frustrating experiences like other users have shared. Probably it was the case with .Net 7. Of course some things take time but almost all help is available online by generous community. Good documentation. And .Net Maui is well thought of in terms of design. My project has about 45 pages, it uses many common controls plus also webview, local notifications, syncing, styles, light dark theme, pictures, collection view etc. Certain places I use SyncFusion controls. The app runs successfully on both Android and iPhone. Haven't launched it yet on the stores as server push notifications is pending.

I see a great future for .Net Maui.

**Reply by MarkusRodler** [reply]
I think you're talking about a different Maui...

**Reply by JohnHDev** [reply]
.NET Maui is looking MUCH better now, we have started migrating a number of apps and it is going well so far. Biggest issue was 3rd party libraries but we have already migrated away from any that are no longer supported to make the transition as smooth as possible.

**Comment by CoolCoderSuper** [comment]
I will say I had used XF a bit and have also built some apps in Flutter but now I am using MAUI and am really liking it.
Fast, fairly simple but still pretty flexible.
I will say if you got your MAUI impressions a year or two ago you should definitely reevaluate.

**Comment by redradist** [comment]
I think that Avalonia is much better from all point of view ...
See this my comment https://github.com/dotnet/maui/discussions/339#discussioncomment-9922179

**Reply by CoolCoderSuper** [reply]
I agree AvaloniaUI is the way for any kind of crossplatform desktop app.
But for mobile it is still really lacking since you generally need much more native interactions on mobile. For example navigation gestures.

**Reply by redradist** [reply]
For native interaction you could use `Xamarin.AndroidX` wrappers around platform SDK

**Reply by CoolCoderSuper** [reply]
Yes, but now I am writing more platform specific code which I wouldn't be with MAUI.

**Reply by redradist** [reply]
> Yes, but now I am writing more platform specific code which I wouldn't be with MAUI.

Why ? You need platform specific only for example BLE, WiFi ... not UI components

**Reply by Stedy59** [reply]
@redradist - "Performance" I don't use the "built in" controls in Maui other than layouts, label, and edit. I authored my own border, button, collection (recyclerview), carousel (viewpager2) for the sake of performance. Too many levels of "indirection" affect performance... Using Maui for desktop, specifically Windows, at this time, is like taking away 50% of performance AND functionality away from WINUI... which is already only 50% capable of other frameworks such as WPF.

As a Principal Software Architect, I view Maui as a viable cross platform for mobile, and Avalonia, desktop.
